lib/trino-parquet/src/main/java/io/trino/parquet/writer/repdef/RepLevelWriterProviders.java

```diff
@@ -0,0 +1,280 @@
+                    offset += positionsCount;
+                }
+            };
+        }
+    }
+
+    static class ColumnMapRepLevelWriterProvider
+            implements RepLevelWriterProvider
+    {
+        private final ColumnarMap columnarMap;
+        private final int maxRepetitionLevel;
+
+        ColumnMapRepLevelWriterProvider(ColumnarMap columnarMap, int maxRepetitionLevel)
+        {
+            this.columnarMap = requireNonNull(columnarMap, "columnarMap is null");
+            this.maxRepetitionLevel = maxRepetitionLevel;
+        }
+
+        @Override
+        public RepetitionLevelWriter getRepetitionLevelWriter(Optional<RepetitionLevelWriter> nestedWriterOptional, ValuesWriter encoder)
+        {
+            checkArgument(nestedWriterOptional.isPresent(), "nestedWriter should be present for column map repetition level writer");
+            return new RepetitionLevelWriter()
+            {
+                private final RepetitionLevelWriter nestedWriter = nestedWriterOptional.orElseThrow();
+
+                private int offset;
+
+                @Override
+                public void writeRepetitionLevels(int parentLevel)
+                {
+                    writeRepetitionLevels(parentLevel, columnarMap.getPositionCount());
+                }
+
+                @Override
+                public void writeRepetitionLevels(int parentLevel, int positionsCount)
+                {
+                    checkValidPosition(offset, positionsCount, columnarMap.getPositionCount());
+                    if (!columnarMap.mayHaveNull()) {
+                        for (int position = offset; position < offset + positionsCount; position++) {
+                            writeNonNullableLevels(parentLevel, position);
+                        }
+                    }
+                    else {
+                        for (int position = offset; position < offset + positionsCount; position++) {
+                            if (columnarMap.isNull(position)) {
+                                encoder.writeInteger(parentLevel);
+                                continue;
+                            }
+                            writeNonNullableLevels(parentLevel, position);
+                        }
+                    }
+                    offset += positionsCount;
+                }
+
+                private void writeNonNullableLevels(int parentLevel, int position)
+                {
+                    int entryLength = columnarMap.getEntryCount(position);
+                    if (entryLength == 0) {
+                        encoder.writeInteger(parentLevel);
+                    }
+                    else {
+                        nestedWriter.writeRepetitionLevels(parentLevel, 1);
+                        nestedWriter.writeRepetitionLevels(maxRepetitionLevel, entryLength - 1);
+                    }
+                }
+            };
+        }
+    }
+
+    static class ColumnArrayRepLevelWriterProvider
```

Comment:
```
A top ColumnArrayRepLevelWriterProvider, `maxRepetitionLevel == 1`
[...] => [[1, 2], [3, 4], [4, 5]]

B nested ColumnArrayRepLevelWriterProvider, B nested in A, `maxRepetitionLevel == 2`
[1, 2], [3, 4], [4, 5]

top guy    => positionCount == 1
nested guy => positionCount == 3
```

* For the B, `columnarArray` `getPositionCount() == 3`
* B for the example above, will never get `parentLevel == 1`!!!, so there is no way to produce rep levels having `1` in chain